psa: use static key slot buffers to store key material

valeriosetti opened the pull request:
This PR is the relevant part of the PR [9448](https://github.com/Mbed-TLS/mbedtls/pull/9448) that was merged in mainline Mbed TLS in LTS branch "mbedtls-3.6". Changes included in this PR will be automatically included in 3.6.3, so by the time Zephyr's Mbed TLS fork repo is bumbed to that official release, this commit MUST be discarded.

This PR introduces the possibility to use static key slot buffers in the PSA core instead of dynamically allocating them when needed. This helps reducing heap memory usage as well as potentially removing heap management ROM code if heap is not used anywhere else in the Zephyr application.

tomi-font commented:
Can you add a list of the commits that were cherry picked to make this one up? (Or just cherry pick the commits directly without modification.)

tomi-font reviewed (approved):
Checked the diff with the upstream PR.

valeriosetti committed:
psa: use static key slot buffers to store key material

This commit takes only relevant changes of PR
https://github.com/Mbed-TLS/mbedtls/pull/9448 that was merged
in upstream Mbed TLS in the LTS branch "mbedtls-3.6".
Since the original PR was made of several commits, but most
of them were only affecting tests cases (not used in Zephyr),
only changes belonging to the "include" and "library" folders
were included here.

== IMPORTANT ==
Changes introduced in this commit will be automatically
part of Mbed TLS release 3.6.3, so by the time Zephyr's
Mbed TLS fork repo is bumbed to that official release,
this commit MUST be discarded.

This commit introduces the possibility to use static key slot
buffers in the PSA core instead of dynamically allocating them
when needed. This helps reducing heap memory usage as well as
potentially removing heap management ROM code if heap is not
used anywhere else in the Zephyr application.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

dkalowsk commented:
As discussed in the Release meeting today, we want to consider the deprecation of Tinycrypt a release blocker item for 4.0. That would mean Part1, Part2, and the mbedTLS changes would all need to be merged in.

@tomi-font @ithinuel  any objection to this course of action? The point was raised that Tinycrypt has been "just about to be deprecated" for 5 years now, but never seems to make it. We'd like to make it happen now.

tomi-font commented:
> As discussed in the Release meeting today, we want to consider the deprecation of Tinycrypt a release blocker item for 4.0. That would mean Part1, Part2, and the mbedTLS changes would all need to be merged in.
>
> @tomi-font @ithinuel any objection to this course of action? The point was raised that Tinycrypt has been "just about to be deprecated" for 5 years now, but never seems to make it. We'd like to make it happen now.

No objection. I'm happy we proceed this way.

valeriosetti commented:
Since 4.0 is out, can we please wmerge this? In this way I can move https://github.com/zephyrproject-rtos/zephyr/pull/80368 (and https://github.com/zephyrproject-rtos/zephyr/pull/79931 consequently) forward.
